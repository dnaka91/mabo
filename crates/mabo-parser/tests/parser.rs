use std::{
    fmt::{self, Display},
    fs,
    path::{Path, PathBuf},
};

use insta::{assert_snapshot, glob, with_settings};
use miette::{Diagnostic, MietteHandler, MietteHandlerOpts, ReportHandler};
use mabo_parser::Schema;

fn strip_path(path: &Path) -> PathBuf {
    path.strip_prefix(concat!(env!("CARGO_MANIFEST_DIR"), "/tests/inputs"))
        .unwrap()
        .to_owned()
}

#[test]
fn parse_schema() {
    glob!("inputs/*.mabo", |path| {
        let input = fs::read_to_string(path).unwrap();
        let value = Schema::parse(input.as_str(), Some(&strip_path(path))).unwrap();

        with_settings!({
            description => input.trim(),
            omit_expression => true,
        }, {
            assert_snapshot!("parse", format!("{value:#?}"));
            assert_snapshot!("print", value.to_string());
        });
    });
}

#[test]
fn parse_invalid_schema() {
    struct Wrapper<'a>(&'a MietteHandler, &'a dyn Diagnostic);

    impl Display for Wrapper<'_> {
        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
            self.0.debug(self.1, f)
        }
    }

    let handler = MietteHandlerOpts::new()
        .color(false)
        .terminal_links(false)
        .width(120)
        .force_graphical(true)
        .build();

    glob!("inputs/invalid/*.mabo", |path| {
        let input = fs::read_to_string(path).unwrap();
        let value = Schema::parse(input.as_str(), Some(&strip_path(path))).unwrap_err();

        with_settings!({
            description => input.trim(),
            omit_expression => true,
        }, {
            assert_snapshot!("error", Wrapper(&handler, &value).to_string());
        });
    });
}
