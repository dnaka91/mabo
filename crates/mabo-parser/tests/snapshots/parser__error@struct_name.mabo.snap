---
source: crates/mabo-parser/tests/parser.rs
description: "struct sample {\n    value: u32 @1,\n}"
input_file: crates/mabo-parser/tests/inputs/invalid/struct_name.mabo
snapshot_kind: text
---
mabo::parse::struct_def (https://docs.rs/mabo-parser/0.1.0/mabo_parser/error/struct.ParseStructError.html)

  × Failed to parse struct declaration
  ╰─▶ Invalid struct name
   ╭─[invalid/struct_name.mabo:1:1]
 1 │ ╭─▶ struct sample {
 2 │ │       value: u32 @1,
 3 │ ├─▶ }
   · ╰──── In this declaration
   ╰────
  help: Expected struct declaration in the form `❬B❭struct <Name> {...}❬B❭`

Error: mabo::parse::struct_def::invalid_name (https://docs.rs/mabo-parser/0.1.0/mabo_parser/error/enum.ParseStructCause.html#variant.InvalidName)

  × Invalid struct name
   ╭─[invalid/struct_name.mabo:1:8]
 1 │ struct sample {
   ·        ▲
   ·        ╰── Problematic character
 2 │     value: u32 @1,
   ╰────
  help: Struct names must start with an uppercase letter (❬Y❭A-Z❬Y❭), followed by zero or more alphanumeric characters
        (❬Y❭A-Z, a-z, 0-9❬Y❭)
